File: test/test_googleles.jl
using Test: Test
using Zarr: Zarr

# Include the GoogleLES module from utils
include("../utils/GoogleLES.jl")
using .GoogleLES: GoogleLES as GL

Test.@testset "Google Cloud Zarr Access Tests" begin
    @info "Testing lazy metadata load from Google Cloud..."
    
    # Test valid site, month, experiment
    ds = GL.load_zarr_simulation(0, 1, "amip")
    
    # 1. Structure Check
    Test.@test ds !== nothing
    Test.@test typeof(ds) <: Zarr.ZGroup
    
    # 2. Variable Check (These are common SwirlLM features)
    # The dataset is a ZGroup, we can check if it has the required data arrays.
    Test.@test haskey(ds.arrays, "p_ref")
    Test.@test haskey(ds.arrays, "T")
    Test.@test haskey(ds.arrays, "q_t")
    
    # 3. Size Check (The raw arrays should have dimensions)
    # For instance temperature shape is (t, x, y, z) ~ (1, 124, 124, 60) per chunk typically,
    # but let's just assert its rank > 0
    t_arr = ds.arrays["T"]
    Test.@test length(size(t_arr)) > 0
    
    @info "Completed Google Cloud Zarr metadata test successfully!"
end

File: test/test_googleles.py
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from utils import GoogleLES as GL

def test_load_zarr():
    print("Testing lazy metadata load from Google Cloud Zarr...")
    try:
        ds = GL.load_zarr_simulation(0, 1, "amip")
        
        # We only loaded metadata (lazy evaluation)
        # Check that we received an xarray Dataset
        import xarray as xr
        assert isinstance(ds, xr.Dataset), "Did not return an xarray Dataset"
        
        # Check for standard SWIRL-LM fields in the Zarr cube
        assert "T" in ds.variables, "Temperature variable 'T' not found in Zarr cube"
        assert "q_t" in ds.variables, "Total water specific humidity 'q_t' not found"
        print("Success! Zarr metadata loaded.")
        ds.close()
    except Exception as e:
        print(f"Test failed with error: {e}")
        sys.exit(1)

def test_load_statistics():
    print("Testing lazy metadata load from Google Cloud NetCDF stats...")
    try:
        ds = GL.load_statistics()
        if ds is not None:
            # Not strictly guaranteed to exist depending on the exact NC file shape
            # but usually they have specific coordinates
            print(f"Stats dataset successfully connected, found {len(ds.variables)} variables.")
            ds.close()
        else:
            print("Failed to connect to the statistics file (it might require authentication or the filename is different).")
    except Exception as e:
        print(f"Test encountered an exception: {e}")
        # Not failing as this file might not be exactly "cloudbench_statistics.nc"
        pass

if __name__ == "__main__":
    test_load_zarr()
    test_load_statistics()
    
    # Cleanly teardown fsspec connections to avoid dangling aiohttp tasks closing error
    import fsspec
    fsspec.clear_instance_cache()
    
    print("All python GoogleLES tests passed (or handled gracefully).")

